app_test.go

```diff
@@ -2265,3 +2265,72 @@ func TestWhenExitSubCommandWithCodeThenAppQuitUnexpectedly(t *testing.T) {
 		t.Errorf("exitCodeFromOsExiter valeu should be %v, but its value is %v", testCode, exitCodeFromExitErrHandler)
 	}
 }
+
+func TestDuplicateCommandName(t *testing.T) {
+	duplicateCommandName := "hello"
+	os.Args = []string{"say", duplicateCommandName}
+	app := NewApp()
+	app.Name = "say"
+	app.Commands = []Command{
+		{
+			Name: duplicateCommandName,
+			Action: func(c *Context) error {
+				fmt.Println("Hello")
+				return nil
+			},
+		},
+		{
+			Name: duplicateCommandName,
+			Action: func(c *Context) error {
+				fmt.Println("Hello duplicate")
+				return nil
+			},
+		},
+	}
+
+	err := app.Run(os.Args)
+	if err == nil {
+		t.Fatalf("expected to receive error from Run, got none")
+	}
+
+	if err.Error() != fmt.Sprintf("list of commands contains duplicate Name: %s", duplicateCommandName) {
+		t.Errorf("received error did not match expectation")
+	}
+}
+
+func TestDuplicateSubcommandName(t *testing.T) {
+	duplicateSubcommandName := "word"
+	os.Args = []string{"say", "kind", duplicateSubcommandName}
+	app := NewApp()
+	app.Name = "say"
+	app.Commands = []Command{
+		{
+			Name: "kind",
+			Subcommands: []Command{
+				{
+					Name: duplicateSubcommandName,
+					Action: func(c *Context) error {
+						fmt.Println("please")
+						return nil
+					},
+				},
+				{
+					Name: duplicateSubcommandName,
+					Action: func(c *Context) error {
+						fmt.Println("thanks")
+						return nil
+					},
+				},
+			},
+		},
+	}
+
+	err := app.Run(os.Args)
+	if err == nil {
+		t.Fatalf("expected to receive error from Run, got none")
+	}
+
+	if err.Error() != fmt.Sprintf("list of child commands of %s contains duplicate Name: %s", "kind", duplicateSubcommandName) {
+		t.Errorf("received error did not match expectation")
+	}
```

Comment:
⭐⭐ _(seriously consider)_ Consider using table driven tests here, here's a guide for them => https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Reply:
Table driven tests are really awesome :) @lynncyrin